From d778fca06db2dacf59e0304d5988da9c90bf2ae0 Mon Sep 17 00:00:00 2001
From: Niko Tyni <ntyni@debian.org>
Date: Mon, 3 Jun 2013 14:33:48 +0300
Subject: Make perlbug.PL run patchlevel.h through cpp where possible

This makes it possible to use #include directives in patchlevel.h
and get them into perlbug reports.

Bug-Debian: http://bugs.debian.org/710842
Patch-Name: fixes/perlbug-cpp.diff
---
 utils/perlbug.PL |   24 +++++++++++++++++-------
 1 file changed, 17 insertions(+), 7 deletions(-)

diff --git a/utils/perlbug.PL b/utils/perlbug.PL
index bf86670..9bdc071 100644
--- a/utils/perlbug.PL
+++ b/utils/perlbug.PL
@@ -24,12 +24,21 @@ open OUT, ">$file" or die "Can't create $file: $!";
 
 # extract patchlevel.h information
 
-open PATCH_LEVEL, "<" . catfile(updir, "patchlevel.h")
-    or die "Can't open patchlevel.h: $!";
+my $cpp = $Config{cpp};
+my $cppflags = $Config{cppflags};
 
-my $patchlevel_date = (stat PATCH_LEVEL)[9];
+my $patchfile = catfile(updir, "patchlevel.h");
 
-while (<PATCH_LEVEL>) {
+my $in;
+if (!open $in, "$cpp $cppflags $patchfile|") {
+    warn "Warning: can't run cpp ('$cpp $cppflags $patchfile'): $!";
+    open $in, "<$patchfile"
+        or die "Can't open patchlevel.h: $!";
+}
+
+my $patchlevel_date = (stat $patchfile)[9];
+
+while (<$in>) {
     last if $_ =~ /^\s*static\s+(?:const\s+)?char.*?local_patches\[\]\s*=\s*{\s*$/;
 }
 
@@ -38,9 +47,10 @@ if (! defined($_)) {
 }
 
 my @patches;
-while (<PATCH_LEVEL>) {
+while (<$in>) {
     last if /^\s*}/;
-    next if /^\s*#/;  # preprocessor stuff
+    next if /^\s*$/;
+    next if /^\s*#/;  # preprocessor stuff in case cpp failed
     next if /PERL_GIT_UNPUSHED_COMMITS/;    # XXX expand instead
     next if /"uncommitted-changes"/;        # XXX determine if active instead
     chomp;
@@ -52,7 +62,7 @@ while (<PATCH_LEVEL>) {
 my $patch_desc = "'" . join("',\n    '", @patches) . "'";
 my $patch_tags = join "", map /(\S+)/ ? "+$1 " : (), @patches;
 
-close(PATCH_LEVEL) or die "Error closing patchlevel.h: $!";
+close($in) or die "Error closing handle to patchlevel.h: $!";
 
 # TO DO (prehaps): store/embed $Config::config_sh into perlbug. When perlbug is
 # used, compare $Config::config_sh with the stored version. If they differ then
